[lgwebos] Extension for Rule Engine

sprehn opened the pull request:
I would like to start this PR to review and discuss how to extend the new Rule Engine. This is a pilot, since archetypes and examples don't exist yet. 

During the PR review of the lgwebos binding it was agreed to remove the action implementation from the actual binding and separate it out into a second bundle.

This extension will add the ability to use LG WebOS smart TVs within the rules system. 

Open topics:
- Dependencies: How to ensure that this bundle can only be installed if the lgwebos binding is installed as well, or pull the dependency automatically?
- additional changes required to integrate into the framework (e.g. features.xml entry ? )
- Naming conventions
  - How to name these extensions? e.g. "rules". 
  - Is /addons/rule the proper location?
- How to access this from within traditional xtext based rules?
- Will it be possible to provide variables in the parameters to actions?
- ...

kaikreuzer commented:
Thanks for starting on such an "action pilot" - please note that I am currently under extreme pressure on all tasks to get openHAB 2.2 released, so please allow me to pause this until after the release (Dec 18).

sprehn commented:
No problem at all.

kaikreuzer commented:
Thanks for your patience!
Let me start with some initial thoughts on your questions:

> Dependencies: How to ensure that this bundle can only be installed if the lgwebos binding is installed as well, or pull the dependency automatically?

My suggestion would be to include it in the very same Karaf feature as the LGWebOS binding (i.e. it comes as a part of the binding). It would then always be installed with it, but the bundle only gets active if the (new) rule engine is installed as well.

> How to name these extensions? e.g. "rules".

I would stick with the naming from openHAB1, i.e. ["action"](https://github.com/openhab/openhab1-addons/tree/master/bundles). -> addons/action

> How to access this from within traditional xtext based rules?

That's to be worked out indeed. We probably should evolve the [action annotations](https://github.com/eclipse/smarthome/blob/master/bundles/model/org.eclipse.smarthome.model.script/src/org/eclipse/smarthome/model/script/actions/Voice.java) and require actions to define an interface that is annotated with those. We could possibly also introduce a generic action for the DSL rules, which can be called with a variable list of params, i.e. something like `callAction("lgwebos.prompt", "My notification")` - this might be the simplest start and would at least allow to use all "new" actions that exist immediately.

> Will it be possible to provide variables in the parameters to actions?

Variables within the rules for sure. Or do you rather think of placeholders that the action itself should replace by something (e.g. some internal state)? Do you have an example to explain your idea here?

kaikreuzer commented:
@sprehn It looks as you have just screwed your branch - could you try to get it back in shape?
Make sure to never do any merge commits, always rebase on master instead!

sprehn commented:
I did rebase, but know I know where my mistake was.. I did not force push, but follow the recommendation to pull, which caused the merge. But I still have some compilation problems due to the upgrade to latest upstream version.

sprehn commented:
@svilenvul 

I am getting two errors produced by PackageExportsNameCheck. But as you can see the other warnings also seem off. 
they are all lacking the leading „org.openhab“ portion of the project name. and the packages on which PackageExportsNameCheck is complaining are „“, the empty string, and „org“.

Any idea what could be wrong with my repo?


> [ERROR] Code Analysis Tool has found: 
>  2 error(s)! 
>  6 warning(s) 
>  5 info(s)
> [WARNING] .binding.lgwebos/ESH-INF/binding/binding.xml:[3]
> There were whitespace characters used for indentation. Please use tab characters instead
> [WARNING] .binding.lgwebos/ESH-INF/thing/thing-types.xml:[3]
> There were whitespace characters used for indentation. Please use tab characters instead
> [WARNING] .binding.lgwebos/pom.xml:[2]
> There were whitespace characters used for indentation. Please use tab characters instead
> [ERROR] .binding.lgwebos/META-INF/MANIFEST.MF:[0]
> The package  should be marked as "internal" if it is not exported.
> [ERROR] .binding.lgwebos/META-INF/MANIFEST.MF:[0]
> The package org should be marked as "internal" if it is not exported.
> [WARNING] .binding.lgwebos.LGWebOS.xml:[3]
> There were whitespace characters used for indentation. Please use tab characters instead
> [WARNING] .binding.lgwebos.internal.discovery.LGWebOSDiscovery.xml:[3]
> There were whitespace characters used for indentation. Please use tab characters instead
> [WARNING] .binding.lgwebos.internal.LGWebOSHandlerFactory.xml:[3]
>

sprehn commented:
@svilenvul : found the issue. there was a hidden file created by my mac

svilenvul commented:
@sprehn : I am glad it is solved now. I saw you email, but didn't have the time to reply, too busy last week :)

Please if you find some problems with the tool, open an issue directly in the https://github.com/openhab/static-code-analysis, thanks !

wborn commented on addons/action/org.openhab.action.lgwebos/.settings/org.eclipse.ltk.core.refactoring.prefs:
```diff
@@ -0,0 +1,2 @@
+eclipse.preferences.version=1
+org.eclipse.ltk.core.refactoring.enable.project.refactoring.history=false
```
The default project settings should suffice so this file can be removed.

wborn commented on addons/action/org.openhab.action.lgwebos/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,32 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-SymbolicName: org.openhab.action.lgwebos;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.3.0.qualifier
```
2.4.0.qualifier

wborn commented on addons/action/org.openhab.action.lgwebos/pom.xml:
```diff
@@ -0,0 +1,19 @@
+  <parent>
+    <groupId>org.openhab.action</groupId>
+    <artifactId>pom</artifactId>
+    <version>2.3.0-SNAPSHOT</version>
```
2.4.0-SNAPSHOT

wborn commented on addons/action/org.openhab.action.lgwebos/pom.xml:
```diff
@@ -0,0 +1,19 @@
+
+  <groupId>org.openhab.action</groupId>
+  <artifactId>org.openhab.action.lgwebos</artifactId>
+  <version>2.3.0-SNAPSHOT</version>
```
2.4.0-SNAPSHOT

wborn commented on addons/action/org.openhab.action.lgwebos/src/main/java/org/openhab/action/lgwebos/handler/AppActionHandler.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
```
Please update all headers to 2018 e.g. by executing the `mvn license:format` at the root of the repository.

wborn commented on addons/action/pom.xml:
```diff
@@ -0,0 +1,22 @@
+  <parent>
+    <groupId>org.openhab.addons</groupId>
+    <artifactId>pom</artifactId>
+    <version>2.3.0-SNAPSHOT</version>
```
2.4.0-SNAPSHOT

wborn commented on addons/pom.xml:
```diff
@@ -18,6 +18,7 @@
     <module>binding</module>
     <module>io</module>
     <module>ui</module>
+    <module>action</module>
```
Is there a reason the module list is no longer sorted?
If so, please document the reason in a comment or otherwise just sort the list. :-)

kaikreuzer commented:
> Since time has passed

Yeah, @sprehn, sorry again for keeping you waiting, but it isn't forgotten. It is actually as one of the next things on the [ESH backlog](https://github.com/eclipse/smarthome/projects/3) and @sjka and myself wanted to dive into it soon.

sprehn commented:
@wborn addressed your review comments and rebased to version 2.4

kaikreuzer commented:
Just ftr: @triller-telekom is currently intensively working on ESH framework features that will make such implementations much simpler - using LGWebOS as a first candidate to validate that approach. I hope we get an update from him on this here soon!

sprehn commented:
Hi

 any update on this?

kaikreuzer commented:
Oh yes! Just today we have merged the very first "binding action" with https://github.com/eclipse/smarthome/pull/6579!

The framework feature has been made available with https://github.com/eclipse/smarthome/pull/6398, which is included in todays ESH stable, which is now available in openHAB.

You can find some documentation about it at https://github.com/eclipse/smarthome/blob/master/bundles/automation/org.eclipse.smarthome.automation.module.core/README.md.

sprehn commented:
Does it mean I should add the actions to my original binding bundle and annotate classes in there?

kaikreuzer commented:
Yes, I think it makes sense to regard such actions in openHAB as an integral part of the binding and thus it is imho fine to have them in the same bundle (while keeping them in a separate package).

sprehn commented:
binding is now amended and I tried to play with it, but I am not sure if the action is properly picked up by the new rule engine. 

If I create a new rule under "then" I see no entries that relate to this binding. Only "enable or disable rules", "execute a given script", etc.. 

Where should I be able to see those actions?

triller-telekom commented:
@sprehn The actions should be exactly where you are looking for them.

Maybe you forgot to override the `getServices()` method in your thinghandler like its done here?

https://github.com/eclipse/smarthome/blob/cc69446f81a9def71a83b276eb90fd02e5ad5f60/bundles/test/org.eclipse.smarthome.magic/src/main/java/org/eclipse/smarthome/magic/binding/handler/MagicActionModuleThingHandler.java#L59-L62

sprehn commented:
@triller-telekom I have implemented getServices() just like that and I also see it is executed and the handler set in the ActionService, when I debug.

Not sure where else to look.

J-N-K commented:
I don‘t think it is somewhere within your binding. I‘m seeing the exact same issue in the DMX binding and I can‘t the actions from the MQTT binding (which seem to work for others).

sprehn commented:
Thx @J-N-K, 
@triller-telekom I set break points in 
AnnotatedActionModuleTypeProvider::addActionProvider and
AnnotationActionModuleTypeHelper::parseAnnotations

they are never executed, even though my IDE's debug configuration includes org.eclipse.smarthome.automation.module.core

maybe we are missing sth in that configuration?

triller-telekom commented:
I guess this might be missing... https://github.com/eclipse/smarthome/pull/6619

kaikreuzer commented:
> I set break points in `AnnotatedActionModuleTypeProvider` they are never executed

This is no surprise as the class that is getting called instead is `AnnotatedThingActionModuleTypeProvider`.

Might you be missing the `org.eclipse.smarthome.automation.thingsupport`bundle in your launch config? That's the one that does the magic.

sprehn commented:
Indeed, adding this bundle did the trick.

sprehn commented:
I am getting several NullPointExceptions in the log.
AnnotatedThingActionModuleTypeProvider
line 66: String thingUID = annotatedThingActions.getThingHandler().getThing().getUID().getAsString();

This is called a couple of times before the ThingHandler is set in my ActionService. After it is set it works.

Could it be that a check for null is missing here?

sprehn commented:
The labels are still untranslated and showing like  "@text/actionLaunchApplicationWithParamLabel"
I created a properties file with english translation. What do I need to do to make the application aware of it? It is named lgwebos_en.properties

triller-telekom commented:
regarding the NPE: I wonder how this can happen, because in

https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java#L155

the handler is set **before** the service is activated, i.e. this occurs before the service can be added in the `AnnotatedThingActionModuleTypeProvider`.

Can you provide some more information why this happens for you?

Regading the translation: It should happen automagically... One example how its done and working is in

https://github.com/eclipse/smarthome/blob/master/bundles/test/org.eclipse.smarthome.magic/src/main/java/org/eclipse/smarthome/magic/binding/internal/automation/modules/MagicMultiServiceMultiActions.java#L57

Is your properties file in the i18n directory inside the ESH-INF directory?

J-N-K commented:
@kaikreuzer thanks, manually adding thingsupport solved it for me (at least in the IDE). But shouldn‘t OpenHAB S1447 include that? What is the correct repo to make an issue?

sprehn commented:
@triller-telekom: it seems in AnnotatedThingActionModuleTypeProvider
```
@Reference(policy = ReferencePolicy.DYNAMIC, cardinality = ReferenceCardinality.MULTIPLE)
public void addAnnotatedThingActions(AnnotatedActionThingHandlerService annotatedThingActions,
```
gets called multiple times and a couple of times before BaseThingHandlerFactory::registerHandler(Thing thing) gets called 

translations, yes I see no difference at the moment.

sprehn commented:
@triller-telekom I upgrade my binding according to the latest changes. And NPE seems to be gone.
I am still facing two issues. 

a) The values from I18N properties files don't get applied. So I still see "@text.." in PaperUI.
b) When I configure a rule it seems like the parameter of actions are not being picked up. I can only give it a name, description and pick a thing. Or should I look elsewhere to set parameter values?

sprehn commented:
Just found the solution to (a). Still looking for a solution for (b).

triller-telekom commented:
@sprehn I guess your solution to a) was that you renamed your default file from "lgwebos_en.properties" to "lgwebos.properties", right? Just for everyone to document...

Regarding your problem b): It highly depends on what you understand by **parameters** of the action. You have to distinguish between **configuration parameters** and **input parameters**. The later one you can specify via @ActionInput etc. And those parameters "flow" from one automation module to the next one, i.e. each module in your rule can write anything to its output values, which can then be consumed by a following module as its input.

If you want to configure an action, we have defined that these configuration parameters are already defined within the selected thing. However, this is what we at first designed before I implemented it. Thinking about it again, we might lack the possibility of setting a configuration on the action which does not affect the thing at all and thus should not become a thing configuration parameter... We might have to extend the API so we can specify a configuration URI on the action. @kaikreuzer What is your opinion on this? I guess you also stumbled across this limitation recently, right?

sprehn commented:
Yes exactly. renamed the file without any locale specifics and it worked as default 

re parameters. my user case is to open a url in the smart tv’s browser. or start an app on the smart tv. show a small message with custom icon. for all of these i need to be able to supply additional parameters. 

--
Sebastian Prehn

> Am 17.12.2018 um 09:42 schrieb Stefan Triller <notifications@github.com>:
> 
> @sprehn I guess your solution to a) was that you renamed your default file from "lgwebos_en.properties" to "lgwebos.properties", right? Just for everyone to document...
> 
> Regarding your problem b): It highly depends on what you understand by parameters of the action. You have to distinguish between configuration parameters and input parameters. The later one you can specify via @ActionInput etc. And those parameters "flow" from one automation module to the next one, i.e. each module in your rule can write anything to its output values, which can then be consumed by a following module as its input.
> 
> If you want to configure an action, we have defined that these configuration parameters are already defined within the selected thing. However, this is what we at first designed before I implemented it. Thinking about it again, we might lack the possibility of setting a configuration on the action which does not affect the thing at all and thus should not become a thing configuration parameter... We might have to extend the API so we can specify a configuration URI on the action. @kaikreuzer What is your opinion on this? I guess you also stumbled across this limitation recently, right?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

sprehn commented:
Hi,
would be great to hear back on the strategic decision on how to handle parameters when calling the those actions. 

As mentioned, the community is eagerly waiting to open a browser on a specific URL, or call an application with additional arguments, etc.. from a traditional rule, btw.

I had it all working when the traditional action was still integrated in this bundle. Back then I was advised to remove the action code from the bundle into a second bundle and use the new rule engine for it. I have now merged it back into the binding bundle, but I am afraid that the new rule engine will note deliver what the LG WebOS TV user base is asking for.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/action/ActionService.java:
```diff
@@ -0,0 +1,230 @@
+ * @author Sebastian Prehn - Initial contribution
+ */
+@ThingActionsScope(name = "lgwebos")
+@Component()
```
OK

sprehn commented:
That is great to hear, so I will experiment with this and update the Readme file accordingly. I would also suggest to merge Pull Request #4492 with various other maintenance first. I have rebased this pull request already accordingly.

sprehn commented:
It only works when I import the Actions class and explicitly type the variable:

**import org.openhab.binding.lgwebos.internal.action.LGWebOSActions**
...
val **LGWebOSActions** actions = getActions("lgwebos","lgwebos:WebOSTV:3aab9eea-953b-4272-bdbd-f0cd0ecf4a46")

otherwise I get this error, when I try to call decreaseChannel: 
'decreaseChannel' is not a member of 'org.eclipse.smarthome.core.thing.binding.ThingActions';

Is this expected?

sprehn commented:
And should this action be in an exported package, as in MQTT binding or in an internal one?

kaikreuzer commented:
Can you confirm that 
```
val actions = getActions("lgwebos","lgwebos:WebOSTV:3aab9eea-953b-4272-bdbd-f0cd0ecf4a46") 
```
gives you a non-null value?
And if you then access that instance, you see the "is not a member of" exception?

sprehn commented:
and
`val actions = getActions("lgwebos","lgwebos:WebOSTV:3aab9eea-953b-4272-bdbd-f0cd0ecf4a46")
`
results in:
`'increaseChannel' is not a member of 'org.eclipse.smarthome.core.thing.binding.ThingActions'`

kaikreuzer commented:
Hey @sprehn, it took me a while to figure out what's wrong. I had to compare it in detail with the MQTT action and this is what I found:
1. You obviously need to expose the Actions class in order to have it accessible by the rule engine.
2. The "old" rule engine requires static methods, which are then offered as "member functions" on their first parameter (i.e. ThingAction instances).

I have created https://github.com/sprehn/openhab2-addons/pull/1, which does these two changes and I can successfully call the LGWebOS action from rules now!

sprehn commented:
That's great, thx! Will continue with my documentation and local changes.

sprehn commented:
Should I remove the annotations on all actions, that require parameters? 

I suspect it does not make sense to offer them in the Rule Engine Paper UI as long as the parameter values cannot be defined.
It would mean only increaseChannel and decreaseChannel would remain for the UI. Everything else would be accessible via the classic rules.

kaikreuzer commented:
> Should I remove the annotations on all actions, that require parameters?

No, this is valuable meta-data that might be consumed in future (even if the current Paper UI does not support it).

sprehn commented:
Yes, agree - but these actions will be offered in the user Interface for selection in the drop down.  Whati if users select it?

sprehn commented:
The build generates one Warning:
[WARNING] .binding.lgwebos/META-INF/MANIFEST.MF:[14]
The exported package `org.openhab.binding.lgwebos.action` is not imported

should this package be imported inthe MANIFEST as well?

kaikreuzer commented:
Yes, please import it as well to make the warning disappear. I missed that, sorry.

sprehn commented:
hm, still get the same behavior in all combinations. 

`Rule 'ChannelUpDown': Could not cast org.openhab.binding.lgwebos.action.LGWebOSActions@5295c4cb to void;`

kaikreuzer commented:
You must not cast. Here's my rule that works:
```
rule "test"
when
    System started
then
    val actions = getActions("lgwebos", "lgwebos:WebOSTV:6f8bc4ca")
    logInfo("LGWebOS", "Action is " + actions)
    actions.decreaseChannel()
end
```

sprehn commented:
Could this be a bug? 

I am getting a NPE in DefaultScriptScopeProvider line 125:
java.lang.NullPointerException: null
	at org.eclipse.smarthome.automation.module.script.defaultscope.internal.DefaultScriptScopeProvider.addThingActions(DefaultScriptScopeProvider.java:125) ~[?:?]

could it be that thingActions is dereferenced during dependency injection before it is initialized during activate?

When I look at the class:
```
 @Reference(policy = ReferencePolicy.DYNAMIC, cardinality = ReferenceCardinality.MULTIPLE)
    void addThingActions(ThingActions thingActions) {
        this.thingActions.addThingActions(thingActions); // <-- NPE here
        elements.put(thingActions.getClass().getSimpleName(), thingActions.getClass());
    }

    @Activate
    protected void activate() {
        busEvent = new ScriptBusEvent(itemRegistry, eventPublisher);
        thingActions = new ScriptThingActions(thingRegistry) // <-- thingActions initialized here
```

kaikreuzer commented:
As it is a dynamic dependency, `addThingActions` can be called before `activate()`, so this is indeed a bug.
If the component isn't activated yet, the method should simply store the thingActions instances and feed them into `this.thingActions` upon activation. Fancy to create a PR with a fix?

sprehn commented:
#6819 to fix the NPE

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/META-INF/MANIFEST.MF:
```diff
@@ -22,6 +22,7 @@ Import-Package:
  org.apache.http.message,
  org.apache.http.protocol,
  org.eclipse.jdt.annotation;resolution:=optional,
+ org.eclipse.smarthome.automation.annotation,
```
you should make this an optional import, so that the binding does not depend on the presence of the new rule engine.

sprehn replied:
Done

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/README.md:
```diff
@@ -30,7 +31,9 @@ If LocalIP is not set, the binding will use openHAB's primary IP address, which
-The binding broadcasts a search message via UDP on the network.
+The binding broadcasts a search message via UDP on the network in order to discover and monitor availability of the TV.
+
+Please note, that if you are runnin openHAB in a docker container you need to use macvlan or host networking for this binding to work.
```
typo: runnin

sprehn replied:
Done

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/action/Application.java~:
```diff
@@ -0,0 +1,42 @@
+/**
```
This file seems to be an accident...

sprehn replied:
yes

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/action/LGWebOSActions.java~:
```diff
@@ -0,0 +1,384 @@
+/**
```
This file seems to be an accident...

sprehn replied:
oh yes, accidentally committed those emacs temp files

sprehn commented:
In the community we are waiting to test the new feature in action, but it seems like the latest snapshot builds for 2.5 don't include the change yet. I just tested using this procedure:

```
docker stop openhab-2.5 && \
docker rm openhab-2.5 && \
docker pull openhab/openhab:2.5.0-snapshot-amd64-debian && \
sudo rm -rf /opt/openhab-exp/userdata/cache && \
sudo rm -rf /opt/openhab-exp/userdata/tmp && \
sudo rm -rf /opt/openhab-exp/userdata/logs/* && \
docker run \
        --name openhab-2.5 \
         --tty \
        -v /etc/localtime:/etc/localtime:ro \
        -v /etc/timezone:/etc/timezone:ro \
        -v /opt/openhab-exp/conf:/openhab/conf \
        -v /opt/openhab-exp/userdata:/openhab/userdata \
        -v /opt/openhab-exp/addons:/openhab/addons \
        -e USER_ID=998 \
        -e GROUP_ID=995 \
        --net=dockernet \
        -d \
        --restart=always \
openhab/openhab:2.5.0-snapshot-amd64-debian

```


```
openhab> bundle:list | grep LG
223 │ Active   │  80 │ 2.5.0.201901122059     │ LG webOS Binding

```

so it is a recent build I assume, but the behavior is still that of before this PR and the other PR #4492, which introduced caching.

Is it just a matter of time, or anything necessary to get this added to the snapshot builds?

kaikreuzer commented:
I would expect that the changes are included, you can see here on the build server that the classes have been compiled during the build: https://ci.openhab.org/job/openHAB2-Bundles/ws/openhab2-addons/addons/binding/org.openhab.binding.lgwebos/target/classes/org/openhab/binding/lgwebos/action/

Can you check your installed jar file (under `userdata/tmp/mvn/org/openhab/binding`) whether it includes those classes?

sprehn commented:
Yes, indeed they are missing, still an older version of the binding, even though everything under userdata/tmp was deleted and was freshly recreated.

sprehn commented:
https://ci.openhab.org/job/openHAB2-Bundles/534/artifact/openhab2-addons/addons/binding/org.openhab.binding.lgwebos/target/

the latest addons build looks good and includes the classes.

sprehn commented:
Well, it is working now. 👍
